src/nupic/utils/SlidingWindow.hpp

```diff
@@ -0,0 +1,202 @@
+        */
+        std::vector<T> getLinearizedData() const;
+
+        bool operator==(const SlidingWindow& r2) const;
+        bool operator!=(const SlidingWindow& r2) const;
+        /** operator[] provides fast access to the elements indexed relatively
+          to the oldest element. So slidingWindow[0] returns oldest element, 
+          slidingWindow[size()] returns the newest. 
+        :param UInt index - index/offset from the oldest element, values 0..size()
+        :return T - i-th oldest value in the buffer
+        :throws 0<=index<=size()
+        */ 
+        T& operator[](UInt index);
+        const T& operator[](UInt index) const;
+
+
+      private:
+        std::vector<T> buffer_;
+        UInt idxNext_;
+}; 
+}} //end ns
+
+/// IMPLEMENTATION
+#include <cmath>
+
+using nupic::util::SlidingWindow;
+
+template<class T>
+SlidingWindow<T>::SlidingWindow(nupic::UInt maxCapacity) :
+  maxCapacity(maxCapacity) {
+  NTA_CHECK(maxCapacity > 0);
+  buffer_.reserve(maxCapacity);
+  idxNext_ = 0;
+}
+
+
+template<class T>
+size_t SlidingWindow<T>::size() const {
+  NTA_ASSERT(buffer_.size() <= maxCapacity);
+  return buffer_.size();
+}
+
+
+template<class T>
+void SlidingWindow<T>::append(T newValue) {
+  if(size() < maxCapacity) {
+    buffer_.push_back(newValue);
+  } else {
+  buffer_[idxNext_] = newValue;
+  }
+  //the assignment must be out of the [] above, so not [idxNext_++%maxCap],
+  // because we want to store the value %maxCap, not only ++
+  idxNext_ = (idxNext_ +1 ) %maxCapacity;
+}
+
+
+template<class T>
+bool SlidingWindow<T>::append(T newValue, T& droppedValue) {
+  //only in this case we drop oldest; this happens always after 
+  //first maxCap steps ; must be checked before append()
+  bool isFull = (buffer_.size()==maxCapacity); 
+  if(isFull) { 
+    droppedValue = buffer_[idxNext_]; 
+  }
+  append(newValue);
+  return isFull;
+}
+
+
+template<class T>
+const std::vector<T>& SlidingWindow<T>::getData() const {
+  return buffer_; 
+}
+
+
+template<class T>
+std::vector<T> SlidingWindow<T>::getLinearizedData() const {
+  std::vector<T> lin;
+  lin.reserve(buffer_.size());
+
+  //insert the "older" part at the beginning
+  lin.insert(std::begin(lin), std::begin(buffer_) + idxNext_, std::end(buffer_));
+  //append the "newer" part to the end of the constructed vect
+  lin.insert(std::end(lin), std::begin(buffer_), std::begin(buffer_) + idxNext_);
+ return lin;
+}
+
+
+template<class T>
+bool SlidingWindow<T>::operator==(const SlidingWindow& r2) const {
+  return ((this->size() == r2.size()) && (this->maxCapacity == r2.maxCapacity) && 
+    (this->getData()== r2.getData()) && getLinearizedData() == r2.getLinearizedData()); 
+  //FIXME review the ==, on my machine it randomly passes/fails the test!
+}
+
+
+template<class T>
+bool SlidingWindow<T>::operator!=(const SlidingWindow& r2) const {
+  return !operator==(r2);
+}
+
+
+template<class T> 
+T& SlidingWindow<T>::operator[](nupic::UInt index) {
```

Comment:
I would like to mark this `T& [] const` but then I get conflict with `const T& [] const`, any way to do that?